### PR TITLE
Async DHCPv4 client

### LINKED ISSUE #89: DHCPv4 async client
Similar to the dhcpv6 we may want to add the v4 asynchronous client

### PR DESCRIPTION
**Pull request by earlgreyz**
A prototype of DHCPv4 Async client. Fixes #89.

### CODE REVIEW

**Comment by earlgreyz** [comment]
The implemented client is for purpose of unicast dhcp requests. Won't work with broadcast but in that case we probably don't need async client anyway.

File: dhcpv4/async/client.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+)
+
+// Default ports
+const (
+	DefaultServerPort = 67
+	DefaultClientPort = 68
+)
+
+// Client implements an asynchronous DHCPv4 client
+// It doesn't use the broadcast socket! Which means it should be used only when
```
Comment (insomniacslk) [inline dhcpv4/async/client.go]: can you open an issue for the broadcast implementation, and link it here? I think it's something we may still want in the future

File: dhcpv4/async/client.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+// NewClient creates an asynchronous client
+func NewClient() *Client {
+	return &Client{
+		ReadTimeout:  dhcpv4.DefaultReadTimeout,
+		WriteTimeout: dhcpv4.DefaultWriteTimeout,
+	}
+}
+
+// Open starts the client
```
Comment (insomniacslk) [inline dhcpv4/async/client.go]: can you add a very brief mention on the meaning of buffer size?

File: dhcpv4/async/client.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+	close(c.sendQueue)
+	close(c.receiveQueue)
+	close(c.errors)
+
+	c.connection.Close()
+}
+
+// Errors returns a channel where runtime errors are posted
+func (c *Client) Errors() <-chan error {
```
Comment (insomniacslk) [inline dhcpv4/async/client.go]: would it make sense to associate errors to transaction IDs? If so, fee free to make a separate PR or open an issue for that

File: dhcpv4/async/client.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+func (c *Client) receive(_ *dhcpv4.DHCPv4) {
+	var (
+		oobdata  = []byte{}
+		received *dhcpv4.DHCPv4
+	)
+
+	c.connection.SetReadDeadline(time.Now().Add(c.ReadTimeout))
+	for {
```
Comment (insomniacslk) [inline dhcpv4/async/client.go]: probably worth deduplicating this code from the synchronous client

File: dhcpv4/async/client.go (hunk 3 above)
Comment (earlgreyz) [reply dhcpv4/async/client.go]: The errors associated with transaction IDs are not posted to this channel, they are `Rejected` in a promise for a given request. These are the errors we were unable to associate with any of the known promises.
